tests/plugins/file_writing_detector.py

```diff
@@ -0,0 +1,27 @@
+from typing import List
+
+from ophyd import Component, Device, Signal
+
+from blueapi.plugins.data_writing import DataCollectionProvider
+
+
+class FakeFileWritingDetector(Device):
+    image_count: Signal = Component(Signal, value=0, kind="hinted")
+    collection_number: Signal = Component(Signal, value=0, kind="config")
+
+    _provider: DataCollectionProvider
+
+    def __init__(self, name: str, provider: DataCollectionProvider, **kwargs):
+        super().__init__(name=name, **kwargs)
+        self.stage_sigs[self.image_count] = 0
+        self._provider = provider
+
+    def trigger(self, *args, **kwargs):
+        return self.image_count.set(self.image_count.get() + 1)
+
+    def stage(self) -> List[object]:
+        collection = self._provider.current_data_collection
+        if collection is None:
+            raise Exception("No active collection")
```

Comment:
This seems pretty readable to me, requires explicit knowledge of a DataCollectionProvider when making the class, but that makes it obvious where the collection number is coming from

Reply:
From discussion, happy with this because it is theoretically easy to strip out later, as long as:

- We keep in mind that we want to strip it out later
- We make an interface or ABC for the singleton in ophyd-async
- We make a DLS-specific implementation in blueapi
- We maintain a parallel implementation in ophyd-async for offline use, that just writes to a preconfigured directory or similar
- We inject references to the singleton into each device